File: leetcode/medium/min_path_sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        M, N = len(grid), len(grid[0])
        dp = [[float('inf')] * (N+1) for _ in range(2)]
        dp[0][1] = 0
        for r in range(1, M+1):
            t = r % 2   # current row
            nt = not t  # previous row
            for c in range(1, N+1):
                x = grid[r-1][c-1]
                dp[t][c] = min(dp[nt][c], dp[t][c-1]) + x
        return dp[M%2][N]

File: leetcode/medium/minPathSum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int M = grid.size();
        int N = grid[0].size();
        vector<vector<int>> dp(2, vector<int>(N+1, INT_MAX));
        dp[0][1] = 0;
        for (int r = 1; r <= M; ++r) {
            int t = r % 2;
            for (int c = 1; c <= N; ++c) {
                int val = grid[r-1][c-1];
                dp[t][c] = val + min(dp[!t][c], dp[t][c-1]);
            }
        }
        return dp[M%2][N];
    }
};
